### PR TITLE
fix timewarp query results when using timezones and crossing DST transitions

### LINKED ISSUE #5156: Timewarp and Timezones
Queries using both `timewarp` post processing and a `PeriodGranularity` specifying a `timezone` do not correctly offset dates if a daylight savings time transition occurs between the source data interval and the query interval. For example, a query of the form

```
{
      "queryType": "timeseries",
      "dataSource": "some_datasource",
      "intervals": "2017-12-04T08Z/2017-12-011T08Z",
      "granularity": {
        "type": "period",
        "period": "P1D",
        "timeZone": "America/Los_Angeles"
      },
      "context": {
        "postProcessing": {
          "type": "timewarp",
          "dataInterval": "2015-06-22/2015-07-20",
          "period": "P1D",
          "origin": "2015-07-20"
        },
        ...
```
returns results that are shifted by an hour
```
[ {
  "timestamp" : "2017-12-03T23:00:00.000-08:00",
  "result" : ...
}, {
  "timestamp" : "2017-12-04T23:00:00.000-08:00",
  "result" : ...
}
...
```
Currently, `TimewarpOperator` is unaware of the timezone and thus any daylight savings time shifts that might occur and should be taken into account when calculating the offset.

### PR DESCRIPTION
**Pull request by clintropolis**
Fixes #5156 

changes:
* `TimewarpOperator` will now compensate for daylight savings time shifts between date translation ranges for queries using a `PeriodGranularity` with a timezone defined
* The `Query` interface has been expanded with 2 new methods, `getGranularity` and `getTimezone`.
* `BaseQuery` now has a `Granularity` and a new constructor that takes an additional argument to set it. When using the original constructor, granularity will be set to `Granularities.ALL`. 
* The `getTimezone` method in `BaseQuery` will return the `PeriodGranularity` timezone if it is used, else it will default to `DateTimeZone.UTC`. 
* `GroupByQuery`, `SearchQuery`, `SelectQuery`, `TimeseriesQuery`, and `TopNQuery` all individually had a granularity, so this change to `Query` and `BaseQuery` cuts down on some duplicate code in addition providing a mechanism for `TimewarpOperator` (and anything else) that needs to be aware of query granularity or timezone. 
* The changes to `BaseQuery` result in serialization differences for `DataSourceMetadataQuery`, `SegmentMetadataQuery`, `ScanQuery`, and `TimeBoundaryQuery`, which previously did not have a granularity, but are now set to `Granularities.ALL`.

### CODE REVIEW

**Comment by leventov** [comment]
Labelling `Design Review` because this PR introduces a new `@ExtensionPoint`.

File: processing/src/main/java/io/druid/query/groupby/GroupByQuery.java [hunk 1]
```diff
@@ -171,15 +170,16 @@ private GroupByQuery(
...
+    super(dataSource, querySegmentSpec, false, context, granularity);
 
     this.virtualColumns = VirtualColumns.nullToEmpty(virtualColumns);
     this.dimFilter = dimFilter;
-    this.granularity = granularity;
     this.dimensions = dimensions == null ? ImmutableList.of() : dimensions;
     for (DimensionSpec spec : this.dimensions) {
       Preconditions.checkArgument(spec != null, "dimensions has null DimensionSpec");
     }
+    Preconditions.checkNotNull(granularity, "Must specify a granularity");
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/groupby/GroupByQuery.java]: This check should be done in `TimeBucketedQuery`'s constructor

File: processing/src/main/java/io/druid/query/select/SelectQuery.java [hunk 2]
```diff
@@ -65,9 +64,8 @@ public SelectQuery(
       @JsonProperty("context") Map<String, Object> context
   )
   {
-    super(dataSource, querySegmentSpec, descending, context);
+    super(dataSource, querySegmentSpec, descending, context, granularity == null ? Granularities.ALL : granularity);
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/select/SelectQuery.java]: I suggest to extract method `Granularities.nullToAll()`

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java [hunk 3]
```diff
@@ -64,11 +63,10 @@ public TimeseriesQuery(
       @JsonProperty("context") Map<String, Object> context
   )
   {
-    super(dataSource, querySegmentSpec, descending, context);
+    super(dataSource, querySegmentSpec, descending, context, granularity);
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: How do timeseries queries behave with `null` granularity?

File: processing/src/main/java/io/druid/query/TimewarpOperator.java [hunk 4]
```diff
@@ -142,7 +153,7 @@ public T apply(T input)
    *
    * @return the offset between the mapped time and time t
    */
-  protected long computeOffset(final long t)
+  protected long computeOffset(final long t, final DateTimeZone tz)
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/TimewarpOperator.java]: Is it possible to refactor this method? Maybe it's code could be simplified?

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java (hunk 3 above)
Comment (clintropolis) [reply processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: I just tested and it throws a null point exception, moving the precondition check from group by into `TimeBucketedQuery` should fix this.

File: processing/src/main/java/io/druid/query/select/SelectQuery.java (hunk 2 above)
Comment (clintropolis) [reply processing/src/main/java/io/druid/query/select/SelectQuery.java]: :+1:

File: processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java (hunk 3 above)
Comment (clintropolis) [reply processing/src/main/java/io/druid/query/timeseries/TimeseriesQuery.java]: It appears `TimeseriesQueryBuilder` defaults to `Granularities.ALL` which is why all of the tests which do not specify a granularity do not explode.

**Comment by leventov** [comment]
@clintropolis could you please also give this PR a title that looks more sensible in git commit history.

**Comment by drcrallen** [comment]
All queries are time bucketed.

`descending` is already a property in `BaseQuery`

Would it make sense for `BaseQuery` to be modified to accept granularity as a constructor, instead of making another class layer that every query will want to use anyways?

**Comment by clintropolis** [comment]
@drcrallen The reason I added `TimeBucketedQuery` is because `DataSourceMetadataQuery`, `SegmentMetadataQuery`, `ScanQuery`, and `TimeBoundaryQuery` all extend `BaseQuery` but did not have a granularity. If you think it is preferable to make `BaseQuery` always have a granularity, I can merge `TimeBucketedQuery` into it and change these other query types to have an implicit 'All' or 'None', I just wasn't sure if conceptually that made sense (or would have any other unintended side effects).

Thanks for the review!

**Comment by gianm** [comment]
> All queries are time bucketed.

They don't all take a granularity parameter, but if you count "hardcoded 'all' granularity" as "time bucketed" then, okay, sure :)

> If you think it is preferable to make BaseQuery always have a granularity, I can merge TimeBucketedQuery into it and change these other query types to have an implicit 'All' or 'None', I just wasn't sure if conceptually that made sense (or would have any other unintended side effects).

If this is the way you go, it should be 'all', since 'none' really means 'millis'.

**Comment by leventov** [comment]
@clintropolis there is a true error at least with intellij inspections, closing/reopening a pr won't help

**Comment by clintropolis** [comment]
Oops indeed, though they appear unrelated to my changes. Should I fix them? But the tests should pass, I think.

**Comment by clintropolis** [comment]
The inspection indicates that of the `getGranularity` and `getTimezone` methods I added to `BaseQuery`, only `getTimezone` should be on the `Query` interface, which I'm not sure makes a lot of sense without the granularity since it supplies the timezone. Should I rework to either remove both from `Query` interface, or does it make sense for them to be there and ignore the inspection somehow?

**Comment by jon-wei** [comment]
I personally think the inspection complaining about `getGranularity` is excessive, but I guess you could move the two methods to BaseQuery from Query, there aren't any Query implementations that aren't also BaseQuery

**Comment by jon-wei** [comment]
@clintropolis You can also add `@SuppressWarnings("unused")` to getGranularity() to suppress that inspection error, I would favor that over moving the methods to BaseQuery

**Pull request merged**